File: apps/java/src/main/java/chapter10/Classifier.java
package chapter10;

import chapter02.Features;
import java.util.List;

/** fit で学習し、predict でラベルを予測する分類器。 */
public interface Classifier {
  /** 訓練データで学習し、自分自身を返す。 */
  Classifier fit(List<Features> x, List<String> t);

  /** 特徴量ごとのラベルを予測する。 */
  List<String> predict(List<Features> x);
}

File: apps/kotlin/src/main/kotlin/chapter10/Classifier.kt
package chapter10

import chapter01.accuracy
import chapter02.TrainTestSplit
import chapter03.DecisionTree
import org.jetbrains.kotlinx.dataframe.AnyFrame

interface Classifier {
    fun fit(
        x: AnyFrame,
        t: List<String>,
    ): Classifier

    fun predict(x: AnyFrame): List<String>
}

class DecisionTreeClassifier(
    maxDepth: Int? = null,
) : Classifier {
    private val tree = DecisionTree(maxDepth)

    override fun fit(
        x: AnyFrame,
        t: List<String>,
    ): DecisionTreeClassifier {
        tree.fit(x, t)
        return this
    }

    override fun predict(x: AnyFrame): List<String> = tree.predict(x)
}

data class Score(
    val train: Double,
    val test: Double,
)

fun evaluate(
    model: Classifier,
    split: TrainTestSplit<String>,
): Score {
    model.fit(split.xTrain, split.tTrain)
    return Score(
        train = accuracy(model.predict(split.xTrain), split.tTrain),
        test = accuracy(model.predict(split.xTest), split.tTest),
    )
}
